Apply listmedia patch which defines the 'list_media' function

doctorfree opened the pull request:
The [RoonCommandLine](https://gitlab.com/doctorfree/RoonCommandLine) project patches the Python Roon API during installation, defining a `list_media` function which takes a search term and returns matching albums, artists, genres, playlists, or tags. If an exact match is not found a partial match is looked for.

Let me know if these modifications are suitable for inclusion or if there is a better way to integrate this functionality.

pavoni commented:
BTW it also closes https://github.com/pavoni/pyroon/issues/30

pavoni commented on roonapi/roonapi.py:
```diff
@@ -398,6 +398,89 @@ def browse_load(self, opts):
+                load_opts["offset"] += PAGE_SIZE
+            if searched >= total_count and found is None:
+                # Don't log an error here, we will do a substring search
+                # LOGGER.error(
```
Perhaps log as debug - or just delete the commented section.

pavoni commented on roonapi/roonapi.py:
```diff
@@ -398,6 +398,89 @@ def browse_load(self, opts):
+            load_opts["offset"] = 0
+            items = self.browse_load(load_opts)["items"]
+
+        LOGGER.debug("Searching Playlists for %s", searchterm)
```
I may have misunderstood - but is this necessarily searching `playlists`

pavoni commented on roonapi/roonapi.py:
```diff
@@ -444,11 +527,12 @@ def play_media(self, zone_or_output_id, path, action=None):
 
                 load_opts["offset"] += PAGE_SIZE
             if searched >= total_count and found is None:
-                LOGGER.error(
```
I need to log an error for home assistant users trying to debug their searches (otherwise it's really hard to use) - but perhaps we could add an optional parameter to`play_media` - perhaps `report_error=True` and you can then call with `False`?

doctorfree replied:
Thanks for the prompt review. I agree with your suggestions and have committed and pushed changes to address them. The only non-trivial change is this one, adding an optional parameter to play_media for logging an error. My tests today were all successful so I believe my change here is ok but I wanted to get your review and perform some additional testing before I mark this as resolved. Let me know what you think.

pavoni reviewed (changes requested):
This looks really good.

A few small suggestions.

At some point we could refactor  `play_media` and `list_media` since they share a lot of code - but for simplicity let's duplicate the code for now.

If you can make the small changes - I'll merge and release a new version.

Thanks for the contribtion!

doctorfree commented:
Another area to review is my use of the special search term "all". I use the "all" search term to indicate I want a match for all media in that category - e.g. all artists or all albums or all playlists in the library. A potential problem arises if there are artists or albums or tags or whatever that is being searched named "all". The search in list_media will not find an artist named "all" but will return all of the artists in the library. With partial matching as I have implemented it, one might search for "all" thinking it would match the something like the artist "Small Faces" but it wouldn't, it would again return all artists.

Perhaps the special search term that indicates give me all of something should be "__all__" or a string less likely to be used in a search. Although "all" isn't really all that likely to be used, I worry about it.

            if searchterm == "all":

doctorfree commented:
After verifying that the special search term "all" was not exposed at the user level, I changed it to "\_\_all\_\_". I've committed the change and pushed it to this branch for inclusion in any potential merge.

pavoni commented:
The new code is throwing a lint error. We need to restructure these huge methods at some point, but let's just disable the issues for now. You can do this with:

```
    def play_media(self, zone_or_output_id, path, action=None, report_error=True):
        # pylint: disable=too-many-locals,too-many-branches
```

pavoni commented:
> 



> After verifying that the special search term "all" was not exposed at the user level, I changed it to "__all__". I've committed the change and pushed it to this branch for inclusion in any potential merge.

Thanks for pointing this out, I'd missed the `magic string`. What do you think about making "" the indicator that everything should be searched. Don't think there is a downside since it isn't a legitimate search string anyway.

And this is quite elegant in python (because the empty string is falsey). You'd need to switch the if test round to pass lint - so it would look like:

```
            if searchterm:
                for item in items:
                    searched += 1
                    if searchterm in item["title"]:
                        matched.append(item["title"])
            else:
                for item in items:
                    searched += 1
                    matched.append(item["title"])
```

doctorfree commented:
I disabled pylint in play_media as you suggested. Let me know if there are build/lint errors or if this now passes.

I agree we need to refactor some code, combining and simplifying play_media and list_media and maybe other functions. I would prefer to leave the special search term as "\_\_all\_\_" for now but I do like how Python handles the empty string. I will see if I can use your suggested magic search empty string as part of the next effort of improvements. Probably should have written the whole thing in Python from the git-go. Anyway, I believe I have sufficiently addressed concerns (if the code now passes pylint). Everything is working well on my end.

I will be unavailable for the next couple of days. Please feel free to raise other issues and continue to make good suggestions but I will probably not be able to address anything until next week. No hurry on my part so this all can wait till then as far as I am concerned. Thanks for your good suggestions. I leave the question of when to merge up to you. I agree, merge now with what we have and work on combining/simplifying/refactoring later but if you wish to refactor first and merge later I would be ok with that too.

pavoni commented:
Happy to do as you suggest. Will release a new version in the next day or so.

Thanks for the contribution.

pavoni commented:
Now released as `0.0.39 `. Thanks again.
